pyathena/sqlalchemy_athena.py

```diff
@@ -215,7 +217,12 @@ def visit_create_table(self, create):
         return text
 
     def post_create_table(self, table):
-        raw_connection = table.bind.raw_connection()
+        if table.bind:
+            return self._post_create_table_with_bind(table, table.bind.raw_connection())
+        else:
+            return self._post_create_table(table)
+
+    def _post_create_table_with_bind(self, table, raw_connection):
```

Comment:
It is better to change the behavior depending on the value of dialect_option as shown below, instead of separating the methods depending on whether the table is bound or not.
```
    def post_create_table(self, table):
        dialect_opts = (
            table.dialect_options["awsathena"]
            if "awsathena" in table.dialect_options
            else table.dialect_options["awsathena.rest"]
        )

        # TODO Supports orc, avro, json, csv or tsv format
        text = "STORED AS PARQUET\n"

        if dialect_opts["location"]:
            location = dialect_opts["location"]
        else:
            raw_connection = table.bind.raw_connection()
            s3_dir = (
                raw_connection._kwargs["s3_dir"]
                if "s3_dir" in raw_connection._kwargs
                else raw_connection.s3_staging_dir
            )
            schema = table.schema if table.schema else raw_connection.schema_name
            location = "{0}{1}/{2}/".format(s3_dir, schema, table.name)
        text += "LOCATION '{0}'\n".format(location)

        if dialect_opts["compression"]:
            compression = dialect_opts["compression"]
        else:
            raw_connection = table.bind.raw_connection()
            compression = raw_connection._kwargs.get("compression")
        if compression:
            text += "TBLPROPERTIES ('parquet.compress'='{0}')\n".format(
                compression.upper()
            )

        return text
```
The diarect_options will be awsathena or awsathena.rest with the dirver name added. awsathena's settings are given priority.
When specifying a location in driver_options, it should be better not to add the schema or table name to the location, since you can describe the settings individually for each table.

Reply:
#### Regarding the dialect options

Can we agree to ignore `table.dialect_options['awsathena.rest']` ? Here are some reasons why.

The predicate `"awsathena" in table.dialect_options` is tautological (always true). The value of `dialect_options["<dialect name[.driver]>"]` value is the result of the merge of the default values specified in the `construct_arguments` defined on the dialect and the actual arguments passed to the construct. Hence:
```python
dialect_opts = (
            table.dialect_options["awsathena"]
            if "awsathena" in table.dialect_options
            else table.dialect_options["awsathena.rest"]
        )
```
Is whatsoever strictly equivalent to:
```python
dialect_opts = table.dialect_options["awsathena"]
```
So you'd have to choose one or the other, or implement some weird reconciliation logic which, if things were to be used this way, would probably have been implemented in SQLAlchemy's core. It's not, has shown by this example using postgresql:
```python
from pprint import pprint
from sqlalchemy import Column, Index, Integer, MetaData, Table

table = Table('t', MetaData(), Column('c', Integer), postgresql_with_oids=False, schema='s')
pprint(dict(table.dialect_options['postgresql.psycopg2']))
# {'on_commit': None,
#  'with_oids': None,     <--- Not overriden
#  'inherits': None,
#  'tablespace': None,
#  'partition_by': None,
#  'ignore_search_path': False}

pprint(dict(table.dialect_options['postgresql']))
# {'on_commit': None,
#  'with_oids': False,    <--- Overriden
#  'inherits': None,
#  'tablespace': None,
#  'partition_by': None,
#  'ignore_search_path': False}
```
Another reason is that you could not feed "dialect+drive" kwarg to the construct: you cannot pass a keyword argument named `<dialect>+<driver>_<argument>` or `<dialect>.<driver>_<argument>`. It's not valid Python syntax. And use `<dialect>_<drive>_<argument>` Yes technically you could, using a dict and `**`, and sadly it does work (with the same reconciliation issues):
```python
try:
    table = Table('t', MetaData(), Column('c', Integer), postgresql_psycopg2_with_oids=False, schema='s')
except ArgumentError:
    pass

table = Table('t', MetaData(), Column('c', Integer), schema='s', **{'postgresql.psycopg2_with_oids': False})
pprint(dict(table.dialect_options['postgresql']))
# {'on_commit': None,
#  'with_oids': None,   <--- Not overriden
#  'inherits': None,
#  'tablespace': None,
#  'partition_by': None,
#  'ignore_search_path': False}

pprint(dict(table.dialect_options['postgresql.psycopg2'])
Out[16]: 
{'on_commit': None,
 'with_oids': False,      <--- Overriden
 'inherits': None,
 'tablespace': None,
 'partition_by': None,
 'ignore_search_path': False}
```
But that's complete abuse of the language, I would consider that more like a bug than anything. And if you look at SQLAlchemy codebase, all dialect kwargs are defined, on the base dialect class.

Finally, from what I understand driver specific dialects are only meant to handle the peculiarities of the DBAPI level, not higher level stuff like statement compilation.

Reply:
When I actually connect to Athena with `awsathena+rest://`, dialect_options stores only `awsathena.rest` and not `awsathena`. Is this a bug in SQLAlchemy? I'm wondering if the correct behavior is to not be able to get dialect_options from `awsathena` when the connection is actually bound by connecting with `awsathena+rest://`.

Reply:
I am not sure I understand what you mean, because I don't see the link between the dialect and the connection.
Maybe you have an code example you can share?

Reply:
What determines whether a dialect kwargs is stored in `<schema_item>.dialect_options['aswsathena']` or `<schema_item>.dialect_options['aswsathena.rest']` is only the prefix you give to the kwargs you pass to the `SchemaItem` constructor:

```python
table = Table('t', MetaData(), Column('c', Integer), awsathena_location='s3://my_bucket/s/t', schema='s')
# Note: as the dialect is currently defined, this works but raises warning and does not take advantages
# of SQLAlchemy builtin dialect kwarg name validation.
print(table.dialect_options['awsathena.rest']['location'])
# None
print(table.dialect_options['awsathena']['location'])
# s3://my_bucket/s/t

table = Table('t', MetaData(), Column('c', Integer), schema='s', **{'awsathena.rest_location': 's3://my_bucket/s/t'})
# Works and does not raise warning, but is simply wrong: ugly syntax, not portable across drivers.
print(table.dialect_options['awsathena.rest']['location'])
# s3://my_bucket/s/t
print(table.dialect_options['awsathena']['location'])
# raises sqlalchemy.exc.NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:awsathena
# Since we never passed such dialect kwargs SQLAlchemy cannot find. The exception is raised because
# in its best effort to provide a default value for 'location', SQLAlchemy tries to to find and load the dialect
# named `awsathena` that does not exists (only awsathena+rest exists).
```
There is no connection involved here.

More over, if you want code to be portable across the two drivers (PyAthena and PyAthenaJDBC), there is no reason to have to specify the driver as part of the dialect kwargs prefix. Athena supports a certain number of SQL statements, independently of the mean by which you connect to it, send it queries and receive back the answers.

Reply:
I was wondering why I was getting the following error in test.
```
E       sqlalchemy.exc.NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:awsathena
```
I understood the root cause.